### PR TITLE
Add response_code accessor and use it internally.

### PR DESCRIPTION
**Pull request by djberg96**
This adds a response_code accessor to our base model, sets it internally for our methods, and uses it within the poll method.

It also modifies the create method to return a response header object instead of a regular object since it's an asynchronous operation, and makes it consistent with the delete method.

### CODE REVIEW

**Comment by djberg96** [comment]
@bzwei @bronaghs Please take a look at this and let me know what you think. The short version is that we need the response code, which is not included in the headers or the body (is it?). For the `poll` method, as per the docs and Boris' email, we can assume success on a 200 or 201.

@bronaghs This (or something like it) is necessary before I can continue with deleting associated resources.

**Comment by djberg96** [comment]
@bzwei I'm having trouble with one test yet, from TemplateDeployments:

```
1) TemplateDeploymentService instance methods defines a get method
     Failure/Error: @hash = JSON.parse(json)
     
     TypeError:
       no implicit conversion of RSpec::ExampleGroups::TemplateDeploymentService::InstanceMethods into String
     # ./lib/azure/armrest/model/base_model.rb:62:in `initialize'
     # ./lib/azure/armrest/resource_group_based_service.rb:78:in `new'
     # ./lib/azure/armrest/resource_group_based_service.rb:78:in `get'
     # ./spec/template_deployment_service_spec.rb:70:in `block (3 levels) in <top (required)>'
```

Can you take a look? I'm not sure how to fix that.

File: spec/template_deployment_service_spec.rb [hunk 1]
```diff
@@ -8,11 +8,9 @@
 describe "TemplateDeploymentService" do
   before do
     setup_params
-
-    class String
-      def body; self; end
-      def headers; {}; end
-    end
+    allow_any_instance_of(String).to receive(:body).and_return(self)
```
Comment (bzwei) [inline spec/template_deployment_service_spec.rb]: returning `self` caused the spec test failure because `self` is `RSpec::ExampleGroups::...InstanceMethods`. A string is expected from the `body` method.

Comment (djberg96) [inline spec/template_deployment_service_spec.rb]: @bzwei Thanks, fixed.

**Comment by bronaghs** [comment]
@djberg96 - We no longer create the object itself eg a VM? We need this object for instance provisioning...
https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb#L143

**Comment by djberg96** [comment]
@bronaghs We still create the object, it's just that the create method returns the response headers instead of internally making another get call.

I suppose I could keep it backwards compatible, and just attach the response info to the object itself.

**Comment by bronaghs** [comment]
@djberg96 - I realize we are still creating the object, what I meant was we need the object itself to be returned. Can you point me to where it is being returned?

**Comment by bronaghs** [comment]
@djberg96 - I mistyped my comment above which explains your response, I should have said:
"We no longer ***return*** the object itself eg a VM? We need this object for instance provisioning..."

**Comment by bzwei** [comment]
I agree with @djberg96 The create method should always return the object it creates.

**Comment by djberg96** [comment]
@bzwei @bronaghs Ok, the :create method returns an object as it did before. I've removed the WIP label, as I think it's ready to go.

**Comment by bronaghs** [comment]
LGTM, @bzwei ?

File: lib/azure/armrest/armrest_service.rb [hunk 2]
```diff
@@ -166,7 +166,8 @@ def tenants
       # such as create or delete.
       #
       def poll(response)
-        url = response.respond_to?(:azure_asyncoperation) ? response.azure_asyncoperation : response.to_s
+        return 'Succeeded' if [200, 201].include?(response.response_code)
```
Comment (bzwei) [inline lib/azure/armrest/armrest_service.rb]: For my curiosity, what is the response_code if the aync operation is not finished yet?
Also, for the case the response contains `azure_asyncoperation` or `location`, what is the return code when it is still ongoing or finished?

Comment (djberg96) [inline lib/azure/armrest/armrest_service.rb]: If it's not finished yet it returns a 202.

https://httpstatuses.com/202

File: lib/azure/armrest/model/base_model.rb [hunk 3]
```diff
@@ -189,7 +190,9 @@ class ResourceProvider < BaseModel; end
     class Sku < BaseModel; end
     class Usage < BaseModel; end
 
-    class ResponseHeaders < BaseModel; end
+    class ResponseHeaders < BaseModel
+      undef_method :response_headers
```
Comment (bzwei) [inline lib/azure/armrest/model/base_model.rb]: undefine a method in a subclass that exists in base class is not an object orient way. Is there other way around?

Comment (djberg96) [inline lib/azure/armrest/model/base_model.rb]: @bzwei I just thought it was inappropriate for a ResponseHeaders object to have a :response_headers method. It was actually causing some confusion for me in development.

If there's a nicer way, I'm listening.

Comment (bzwei) [inline lib/azure/armrest/model/base_model.rb]: @djberg96 `ResponseHeaders` does not need to be based on the `BaseModel`.

Comment (djberg96) [inline lib/azure/armrest/model/base_model.rb]: @bzwei Well, I need to be able to create a model object from the response.headers output from RestClient. What do you suggest then?

Comment (djberg96) [inline lib/azure/armrest/model/base_model.rb]: Not really, since I need the automatic method generation.

I think down the road we should actually create an ExtendedModel subclass that adds additional accessors like :response_headers, plus any other tweaks we need. Most models would then subclass ExtendedModel, while something like ResponseHeaders would just subclass BaseModel. But, I would ask that be a separate PR.

**Comment by bzwei** [comment]
OK. Subject to further PR refactoring about the base class
